### PR TITLE
Make rpc.shutdown()/rpc._wait_all_workers() accept user-specified leader worker

### PR DESCRIPTION
**Pull request by xush6528**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32635 Use the C++ current RpcAgent pointer to eliminate the unnecessary argument passing from Python world
* #32633 Make C++ RpcAgent::defaultRPCAgent() the source of truth of current RPC Agent
* #32624 Make _wait_all_workers() support being called for multiple times
* **#32623 Make rpc.shutdown()/rpc._wait_all_workers() accept user-specified leader worker**

In the current `_wait_all_worker()` implementation, the procesure is as follows,
- A leader is elected.
- Every worker shows the shutdown intention to the leader node.
- After receiving intentions from all workers, the leader broadcasts the shutdown request to all follower nodes.

While in a user use case, (training toolkit), user has a similar implementation that, where
- The leader is specified by user.
- The termination is initialized by the "leader" node, regardless of the follower nodes' progress.

This PR adds the second protocol, with only one extra argument added.
This helps reduce duplicate implementations scattered around in user codebase.

Differential Revision: [D16453494](https://our.internmc.facebook.com/intern/diff/D16453494/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D16453494/)!

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 12dfc623:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=12dfc623c8fca1ebb5158c5698aa6ad4e96006b5).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4342715) pytorch_linux_xenial_cuda10_1_cudnn7_py3_nogpu_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90913566) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90913566))
<details>
<summary>
<code>Jan 26 00:16:21 RuntimeError: test_quantization failed!</code>
</summary>

```
Jan 26 00:16:21 Ran 36 tests in 47.656s 
Jan 26 00:16:21  
Jan 26 00:16:21 FAILED (errors=1, skipped=1) 
Jan 26 00:16:21  
Jan 26 00:16:21 Generating XML reports... 
Jan 26 00:16:21 Traceback (most recent call last): 
Jan 26 00:16:21   File "test/run_test.py", line 456, in <module> 
Jan 26 00:16:21     main() 
Jan 26 00:16:21   File "test/run_test.py", line 449, in main 
Jan 26 00:16:21     raise RuntimeError(message) 
Jan 26 00:16:21 RuntimeError: test_quantization failed! 
Jan 26 00:16:21 + cleanup 
Jan 26 00:16:21 + retcode=1 
Jan 26 00:16:21 + set +x 
Jan 26 00:16:21 =================== sccache compilation log =================== 
Jan 26 00:16:21 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 26 00:16:21 Compile requests                  7 
Jan 26 00:16:21 Compile requests executed         6 
Jan 26 00:16:21 Cache hits                        0 
Jan 26 00:16:21 Cache misses                      6 
Jan 26 00:16:21 Cache timeouts                    0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/distributed/rpc/api.py [hunk 1]
```diff
@@ -123,7 +123,10 @@ def _wait_all_workers():
     assert (
         _ALL_WORKER_NAMES is not None
     ), "`_ALL_WORKER_NAMES` is not initialized for `def _wait_all_workers`."
-    leader_worker_name = sorted(_ALL_WORKER_NAMES)[0]
+    if exit_on_worker_done is not None:
+        leader_worker_name = _to_worker_info(exit_on_worker_done).name
```
Comment (mrshenli) [inline torch/distributed/rpc/api.py]: Any reason for using the name `exit_on_worker_done`? can we just use `leader`?

File: torch/distributed/rpc/api.py [hunk 2]
```diff
@@ -168,7 +173,7 @@ def _wait_all_workers():
 
 
 @_require_initialized
-def shutdown(graceful=True):
+def shutdown(graceful=True, exit_on_worker_done=None):
```
Comment (mrshenli) [inline torch/distributed/rpc/api.py]: I am OK with adding one arg to `_wait_all_workers`, but have a little hesitation in adding it here:

1. It's like we expose the internal termination algorithm to the user API. 
2. Is it true that `exit_on_worker_done` only work with `graceful = True`?
3. What is the implication here if we add elasticity in the future, i.e., allow nodes to join and leave dynamically? (cc @agolynski)

Comment (mrshenli) [reply torch/distributed/rpc/api.py]: Does this actually mean we should make termination algorithm pluggable?

File: torch/distributed/rpc/api.py [hunk 3]
```diff
@@ -137,13 +140,15 @@ def _wait_all_workers():
...
-        )
+        if exit_on_worker_done is None:
+            rpc_sync(
+                leader_worker_name,
+                _on_leader_follower_report_shutdown_intent,
+                args=(self_worker_name,),
+            )
 
-    _SHUTDOWN_PROCEED_SIGNAL.wait()
+    if not is_leader_worker or (is_leader_worker and exit_on_worker_done is None):
```
Comment (zzzwen) [inline torch/distributed/rpc/api.py]: Q: for this case `is_leader_worker and exit_on_worker_done is None`, who will be setting the event `_SHUTDOWN_PROCEED_SIGNAL`

File: torch/distributed/rpc/api.py (hunk 1 above)
Comment (xush6528) [reply torch/distributed/rpc/api.py]: I'll update this to leader.

File: torch/distributed/rpc/api.py (hunk 2 above)
Comment (xush6528) [reply torch/distributed/rpc/api.py]: > 1. It's like we expose the internal termination algorithm to the user API.
> 2. Is it true that `exit_on_worker_done` only work with `graceful = True`?

> Does this actually mean we should make termination algorithm pluggable?

For 1, Yeah, I am making users have to make a choice between 2 algorithms.
For 2, you are right.

I agree that this is not a clean API yet.

Good idea. I am indeed adding a new shutdown algorithm/strategy. 
It's better to make it pluggable, and pass a per-strategy options struct, like what we do for different RPC backends.

Let me try to extract an interface out of it.

**Comment by xush6528** [comment]
No need to add a second algorithm. Will migrate user to the exisintg shutdown algorithm.

**Pull request closed without merging**